tutorials/arvo/cli-tutorial.md

```diff
@@ -0,0 +1,463 @@
+verb`) is called, which allows the agent to print what it is doing.
+
+```hoon
+|_  =bowl:gall
++*  this  .
+    def   ~(. (default-agent this %|) bowl)
+    des   ~(. (default:shoe this command) bowl)
+::
+```
+This is boilerplate Gall agent core code. We set `def` to be an alias for the part of the
+context where the `default-agent` lives, and set `des`  to be an alias for the part of the
+context where `shoe` library commands live.
+
+Next we begin implementing all of the arms 
+
+```hoon
+++  on-init   on-init:def
+++  on-save   !>(state)
+++  on-load
+  |=  old=vase
+  ^-  (quip card _this)
+  [~ this]
+::
+++  on-poke   on-poke:def
+++  on-watch  on-watch:def
+++  on-leave  on-leave:def
+++  on-peek   on-peek:def
+++  on-agent  on-agent:def
+++  on-arvo   on-arvo:def
+++  on-fail   on-fail:def
+```
+Boilerplate Gall app arms using the minimum implementation found in `def`.
+
+Here begins the implementations of the additional arms required by the
+`(shoe:shoe command)` interface.
+```hoon
+++  command-parser
+  |=  sole-id=@ta
+  ^+  |~(nail *(like [? command]))
+  (cold [& ~] (jest 'demo'))
+```
+`+command-parser` is of central importance - it is what is used to parse user
+input and transform it into `$command`s for the app to execute. Writing a proper
+command parser requires understanding of the Hoon parsing functions found in the
+standard library, but we will cover the bare essentials of parsing here.
+
+`+command-parser` is a gate which takes in a `sole-id=@ta` identifying a `sole`
+session and produces a particular sort of gate which takes in a `nail` (parse
+input) and returns an `edge` (parse output) that we call a parser or `$rule`.
+`+command-parser` produces a specific type of `$rule` whose result is of the
+type `[? command]`. Let us elaborate.
+
+ `nail` is the remainder of a parse given by `[p=hair q=tape]` where the
+`hair=[p=@ud q=@ud]` represent line and column of the current position of the parse and the
+`tape` is the parsing input. An `edge` is a given by `[p=hair q=(unit
+[p=* q=nail])]`. `p` says what we have parsed up to this point. `q` is the
+result of the parse, which is `~` if nothing valid has yet been parsed or a noun
+`p.q` that here will be a value of the `$command` type and `q.q` a `nail` that
+contains results that may still be unparsed. Here the
+`edge` is cast with `*(like [? command])`. `like [? command]`
+produces a generic `edge` which returns a result that is a cell of values of the
+`?` and `command` types - this is what forces `+command-parser` to produce a
+`$rule` that returns a value of type `[? command]`.
+
+Anyways, the gate is
+```hoon
+(cold [& ~] (jest 'demo'))
+```
+`cold` is a parser modifier that accepts a constant noun (here `[& ~]`) and a
+`rule` (`(jest 'demo')`) to produce another `rule`. This `rule` is then a parser that
+produces `[& ~]` whenever the `rule` `(jest 'demo')` triggers. `(jest 'demo')`
+triggers whenever the `tape` in the input `nail` exactly
+matches `demo`.
```

Comment:
```suggestion
`+cold` takes a noun and a parser. The result is a parser that's identical to the parser we passed in, except it always produces our noun (`[& ~]`, matching `[? command]`) if parsing succeeds, instead of whatever the parser itself would produce.
```